src/diffusers/pipelines/versatile_diffusion/pipeline_versatile_diffusion.py

```diff
@@ -0,0 +1,312 @@
+                the, usually at the expense of lower image quality.
+            generator (`torch.Generator`, *optional*):
+                A [torch generator](https://pytorch.org/docs/stable/generated/torch.Generator.html) to make generation
+                deterministic.
+            output_type (`str`, *optional*, defaults to `"pil"`):
+                The output format of the generate image. Choose between
+                [PIL](https://pillow.readthedocs.io/en/stable/): `PIL.Image.Image` or `np.array`.
+            return_dict (`bool`, *optional*):
+                Whether or not to return a [`~pipeline_utils.ImagePipelineOutput`] instead of a plain tuple.
+
+        Returns:
+            [`~pipeline_utils.ImagePipelineOutput`] or `tuple`: [`~pipelines.utils.ImagePipelineOutput`] if
+            `return_dict` is True, otherwise a `tuple. When returning a tuple, the first element is a list with the
+            generated images.
+        """
+        do_classifier_free_guidance = guidance_scale > 1.0
+
+        if height % 8 != 0 or width % 8 != 0:
+            raise ValueError(f"`height` and `width` have to be divisible by 8 but are {height} and {width}.")
+
+        if prompt is not None:
+            if isinstance(prompt, str):
+                batch_size = 1
+            elif isinstance(prompt, list):
+                batch_size = len(prompt)
+            else:
+                raise ValueError(f"`prompt` has to be of type `str` or `list` but is {type(prompt)}")
+
+            condition_embeddings = self._encode_prompt(prompt, do_classifier_free_guidance)
+            prompt_type = "text"
+        elif image_prompt is not None:
+            if isinstance(image_prompt, PIL.Image.Image):
+                batch_size = 1
+            elif isinstance(image_prompt, torch.Tensor):
+                batch_size = image_prompt.shape[0]
+            else:
+                raise ValueError(
+                    f"`image_prompt` has to be of type `PIL.Image.Image` or `torch.Tensor` but is {type(image_prompt)}"
+                )
+
+            condition_embeddings = self._encode_image_prompt(image_prompt, do_classifier_free_guidance)
+            prompt_type = "image"
+        else:
+            raise ValueError("Either `prompt` or `image_prompt` has to be provided.")
+
+        latents = torch.randn(
+            (batch_size, self.image_unet.in_channels, height // 8, width // 8), generator=generator, device=self.device
+        )
+
+        self.scheduler.set_timesteps(num_inference_steps)
+
+        # prepare extra kwargs for the scheduler step, since not all schedulers have the same signature
+        accepts_eta = "eta" in set(inspect.signature(self.scheduler.step).parameters.keys())
+
+        extra_kwargs = {}
+        if accepts_eta:
+            extra_kwargs["eta"] = eta
+
+        with VersatileMixedModel(self.image_unet, self.text_unet, "image", prompt_type) as unet:
```

Comment:
Aiii I'll need a lot of convincing for such an API -> we never use context managers in pipelines and it's very difficult / unintuitive too understand what's going on here IMO

Reply:
Just experimenting with it for now, no worries :) 
The ultimate goal is to ensure that after swapping the layers we swap them back to reuse for another task